clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java

```diff
@@ -0,0 +1,193 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Abstract assignor implementation which uses the following metadata format:
+ *
+ * ProtocolMetadata => Topics TopicPattern MetadataHash
+ *   Topics         => [String]
+ *   TopicPattern   => String
+ *   MetadataHash   => Bytes
+ *
+ * This class handles metadata consistency checking between group members. Custom assignors will
+ * usually extend this class unless they need to provide their own metadata format.
+ */
+public abstract class AbstractPartitionAssignor implements PartitionAssignor<AbstractPartitionAssignor.ConsumerMetadata> {
+    private static final String TOPIC_KEY_NAME = "topics";
+    private static final String TOPIC_PATTERN_KEY_NAME = "topic_pattern";
+    private static final String METADATA_HASH_KEY_NAME = "metadata_hash";
+
+    public static final Schema CONSUMER_METADATA_SCHEMA = new Schema(
+            new Field(TOPIC_KEY_NAME, new ArrayOf(Type.STRING)),
+            new Field(TOPIC_PATTERN_KEY_NAME, Type.STRING),
+            new Field(METADATA_HASH_KEY_NAME, Type.BYTES));
+
+    private static final Type CONSUMER_METADATA_TYPE = new ConsumerMetadataType();
+
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public abstract List<TopicPartition> assign(String consumerId,
+                                                Map<String, Integer> partitionsPerTopic,
+                                                Map<String, ConsumerMetadata> consumerMetadata);
+
+    public Type schema() {
+        return CONSUMER_METADATA_TYPE;
+    }
+
+    @Override
+    public AssignmentResult assign(String consumerId,
+                                   Map<String, ConsumerMetadata> consumers,
+                                   MetadataSnapshot metadataSnapshot) {
+        byte[] topicMetadataHash = metadataSnapshot.hash();
+        Set<String> allSubscribedTopics = new HashSet<>();
+        boolean consistentMetadata = true;
+
+        for (ConsumerMetadata metadata : consumers.values()) {
+            byte[] memberTopicMetadataHash = Utils.toArray(metadata.topicMetadataHash);
+            consistentMetadata = consistentMetadata && Arrays.equals(topicMetadataHash, memberTopicMetadataHash);
+            allSubscribedTopics.addAll(metadata.topics);
+        }
+
+        if (!consistentMetadata) {
+            log.debug("Assignment failed due to inconsistent metadata among consumers");
+            return AssignmentResult.inconsistentMetadata(allSubscribedTopics);
+        }
+
+        Map<String, Integer> partitionsPerTopic = new HashMap<>();
+        for (String topic : allSubscribedTopics) {
+            Integer numPartitions = metadataSnapshot.topicMetadata().get(topic);
+            if (numPartitions != null)
+                partitionsPerTopic.put(topic, numPartitions);
+            else
+                log.debug("Skipping assignment for topic {} since no metadata is available", topic);
```

Comment:
Does this really make sense? For this to occur, we'd need to have everyone agree on the metadata since that was checked above, but then they'd have to be missing metadata on one of those topics. Wouldn't at least the one consumer that included that topic have metadata for it?

Is the goal here that we continue processing the ones we have metadata for so we can at least make progress? If we do this, is there anything that's forcing the metadata to be refreshed (like the inconsistentMetadata result does)? If not, wouldn't this cause us to sometimes knowingly ignore some topics which we might be able to make progress on immediately if we refreshed and rejoined the group?